feat: sbom-convert cli

manifestori opened the pull request:
This PR includes the initial release of `sbom-convert` CLI. 

Spec discussed [here](https://www.notion.so/manifestcyber/Project-Spec-SBOM-Translator-cli-d2011cacbb94467997b8b971e6972e49?pvs=4)

PR is based on the work made by @houdini91 at #1 and on existing work from the go-cli @puerco, considering best practices, tools, and personal opinions. 

Not supported yet:
- config from file
- options in help command 

Several things do need work:
- text for commands description
- Makefile / [Taskfile](https://taskfile.dev/) 
- Improved CI / PR workflows
- cleanup of .gitignore
- more public releases (nix/scoop etc)
- README.md / CONTRIBUTE.md etc.
- Improved error construct and error handling
- Tweaking golangci-lint rules 
- Tests

Notes on implementation:

**Structure**:  we might want to move "convert" to /pkg from /internal as people might want to integrate the full functionality of convert as a library. 

**Interface**: this CLI accepts only `-f` or `--format` and accepted values are `spdx` `spdx-2.2` ... `cyclonedx-1.4` etc.
we do not have a `--spec-version` which might be a better UX. so instead of:

`sbom-convert convert samples/bom.spdx.json --debug -f cyclonedx-1.4 -o output.json`
to
`sbom-convert convert samples/bom.spdx.json --debug -f cyclonedx --spec-version=1.4 -o output.json`

I prefer the former, but we can also support both cases.

**Output**: default behaviour for output is to use stdout if no `-o` parameter is given. however, since we didn't have validations over the file name, we prefer the spec (ie `file.spdx.json`). I put an auto fix for names, adding .cdx or .spdx and .json to file names if needed.  I'm happy to change this behavior and revert to "no autofix" but with proper file name validations (name.[format-ext].[encoding-ext]).

Release & Demo: 
As you might noticed, I created a repo for homebrew and scoop, which will help releasing for mac/windows.
I tested the release pipeline on my fork: `https://github.com/manifest-cyber/sbom-convert/actions` you can look at the release in action. It uses goreleaser to release to Fury:
![image](https://github.com/bom-squad/go-cli/assets/116580393/3743af61-cdd1-4a79-9ade-c2d60bdd49c0)
while also tagging and releasing to Github
![image](https://github.com/bom-squad/go-cli/assets/116580393/779c9323-8913-4b17-86b5-38bf6bf5030f)
With proper releases of [Docker images](https://github.com/manifest-cyber/sbom-convert/pkgs/container/sbom-convert) (no dockerhub yet)
![image](https://github.com/bom-squad/go-cli/assets/116580393/4ac968d5-6fb6-4bf5-adf8-31abe703e1f7)

I know it has a lot of changes in it, I'm happy to answer any questions :) 
🥂

jspeed-meyers commented:
@houdini91

jspeed-meyers commented on .devcontainer/post-create.sh:
```diff
@@ -0,0 +1,93 @@
+    fi
+done
+
+# Install Node tools
```
Does this script need the `node` tools?

manifestori replied:
yes, I'm using it to install @commitlint/cli commitlint/config-conventional

jspeed-meyers commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM alpine:3.18
```
Any reason for using a tag and not a digest?

manifestori replied:
Ill try using `sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1`  (latest 3.18.2)

jspeed-meyers reviewed (approved):
This looks great. An end to end test or two would be nice, but that can wait for a separate PR too. [Caveat: I'm a s*** Go programmer, so you should probably get a review in addition to mine.]

houdini91 commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,201 @@
+  use: buildx
+  build_flag_templates:
+  - "--pull"
+  - "--label=io.artifacthub.package.readme-url=https://raw.githubusercontent.com/{{ .Env.ORG_NAME }}/{{ .ProjectName }}/main/README.md"
```
Awesome, i am going to steal this.

houdini91 commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,201 @@
+
+release:
+ disable: false
+ name_template: 'v{{ .Version }}'
```
if i understand correctly you tag the repo with x.y.z and it creates a version with vX.Y..Z?

manifestori replied:
version is without "v"  (if u do --version) but publicly, tags and releases have it.

houdini91 replied:
Can you elaborate a bit about the standard or spec behind this?

If i understand your saying i will tag `git tag v0.0.1` and the images,git tags and releases artifacts will have v0.0.0 right so its basically only the version command that will have the unprefixed version, if this is the case i have no issue here , resolve the comment.

I just say my two sense on this.
Personally i prefer consistency, looking forward some CDX & SPDX formats allow you to include the tool identification creating them (its probably not required) .
It may be valuable to have consistency with the public artifact versing scheme.
Now after all this blaber i guess we will simply add the `v` if need be..

manifestori replied:
We can streamline that no problem, I'm fine with going for v0.0.0 everywhere. I'll just add `v` in front of the version :D

manifestori replied:
![image](https://github.com/bom-squad/go-cli/assets/116580393/f344ba8d-4221-4488-adf5-3d8cb8365bf5)

houdini91 commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,201 @@
+    flags:
+      - -trimpath
+    ldflags:
+      - -s -w -X {{ .ModulePath }}/cmd/cli.version={{.Version}}
```
Awesome, this should allow the CLI to run `go-cli --version` like before or do we have some other way to get the version from the CLI?

manifestori replied:
ldflags works the same way, so just `--version` (-v is for verbose so only --version)

![image](https://github.com/bom-squad/go-cli/assets/116580393/08de7421-14c5-41d4-ba7d-2a794663bfa6)

houdini91 commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+CMD [ "-h" ]
+
+COPY sbom-convert_*.apk /tmp/
+RUN apk add --no-cache --allow-untrusted /tmp/sbom-convert_*.apk
```
i think goreleaser actually allows you to access the built binary artifact.
i can see how installing a APK can allow better version management on top of the API based image.

We can also go the route of a scrach based image with the single binary, just putting it out there see what people think..

houdini91 commented on cmd/cli/options/root.go:
```diff
@@ -0,0 +1,19 @@
+func (ro *RootOptions) AddFlags(cmd *cobra.Command) {
+	cmd.PersistentFlags().StringVarP(&ro.ConfigPath, "config", "c", "", "path to config file")
+	cmd.PersistentFlags().BoolVarP(&ro.Verbose, "verbose", "v", false, "verbose output")
+	cmd.PersistentFlags().BoolVar(&ro.Debug, "debug", false, "debug output")
```
can we add a `level` flag or at maybe support `-vv`.

manifestori replied:
sure, we can't easily make debug with -vv . btw only place i've seen the usage of -vv (or -vvv) is with syft :) but ill add that and remove the --debug flag so we have -v (info) -vv (debug) -vvv (trace, if supported)

houdini91 commented on cmd/cli/root.go:
```diff
@@ -0,0 +1,94 @@
+	"github.com/bom-squad/go-cli/pkg/log"
+)
+
+var version = "dev"
```
we should consider having a semantic format default version

manifestori replied:
`0.0.0-dev` as the default, GoReleaser sets the version, so it doesn't matter.

houdini91 commented on cmd/cli/root.go:
```diff
@@ -0,0 +1,94 @@
+		Use:     "sbom-convert",
+		Version: version,
+		Short:   "",
+		Long:    ``,
```
Should we add a minimum required arguments?
Also place holder comment to Fill in the long, short and examples field at some point .

manifestori replied:
`Args:              cobra.ExactArgs(1),` 

Yes, adding short/long placeholder + examples.

houdini91 commented on cmd/cli/root.go:
```diff
@@ -0,0 +1,94 @@
+	return nil
+}
+
+func setupLogger(ro *options.RootOptions) error {
```
i suggest we somehow allow the CLI to pass a instance of the logger to the library or some way for the logger to access the global logger, 

I was thinking we should use a interface so the Library can add all  shorts of information to the logger.
but they may be other methods to do this, or simply decide to chose a specific logger to use across the repos.

manifestori replied:
I have raised a concern regarding `protobom` as it should only function as a library without directly requiring a logger or outputting anything to stdout unless specified. However, I am not in favor of including a logger with stdout. 

I concur that it would be best to support an external logger (passed from the CLI) as an interface that defaults to stdout/log.

houdini91 replied:
>  I am not in favor of including a logger with stdout.
Can you elaberate what you mean i don't fully follow.

I guess we can leave the logger out of this PR, lets try and discuss this @puerco @manifestori See if we can present a interface (we need both the cli and library to agree on this).

houdini91 commented on cmd/cli/options/convert.go:
```diff
@@ -0,0 +1,19 @@
+
+// AddFlags adds command line flags for the ConvertOptions struct
+func (o *ConvertOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&o.Format, "format", "f", "", "format string")
```
We should think about default values and description that include that action supported values, I suggest pulling the list of supported formats from the library and populating some flag fields with the information.
EG: https://github.com/scribe-security/protobom/blob/main/pkg/formats/formats.go#L29

We should also consider the URI format defined my our core library.

I suggest supporting this format of a string as well go a long way with people looking at the spec, we can support other aliases as well we may find useful. (in my PR i added a URI flag you can look at).

manifestori replied:
Yeah I know I didn't add the options to doc, but I would add them (manually)

I think supporting a URI is out of scope for this PR.

I will be adding [cyclonedx, cyclonedx-1.3, cyclonedx-1.4, spdx, spdx-2.2, spdx-2.3]

manifestori replied:
The issue with pulling from this https://github.com/scribe-security/protobom/blob/main/pkg/formats/formats.go#L29 is that it uses the URI format, which we need translation for a more friendly alias.

Also, adding a new spec version rarely happens, thinking about cost-effectiveness.

houdini91 replied:
i actually had the same thought in my own PR and tried to tackle it already, your welcome to look in to it and decide if its useful for or not..

* Take a look at a helper function to translate between the URI list. 
[format_group.go](https://github.com/scribe-security/go-cli/blob/suggestion/basecli/internal/config/format_group.go)

* I also added to the options struct has a helper function to simply meld in the flags and spit out the URI needed to access the library. 
[application.go](https://github.com/scribe-security/go-cli/blob/suggestion/basecli/internal/config/application.go#L53)

houdini91 replied:
BY the way note that if we decide to break down the CLI in to two subcommand then the format_group can return slices instead of maps.

manifestori replied:
dully noted. Let's open a new issue for that, sounds like a very good "first good issue" kinda thing; cool but not essential

houdini91 commented on cmd/cli/options/convert.go:
```diff
@@ -0,0 +1,19 @@
+func (o *ConvertOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&o.Format, "format", "f", "", "format string")
+	cmd.Flags().StringVarP(&o.Encoding, "encoding", "e", "", "output encoding")
+	cmd.Flags().StringVarP(&o.OutputPath, "output", "o", "", "path to write the converted SBOM")
```
i am a bit confused about the `--encoding` and `--format` flags ,
i would love to allow the user to only change the version similar to the encoding one.

manifestori replied:
I took the "encoding" idea from your PR. encoding being [json, text, ... ] when format being [spdx, cyclonedx, cyclonedx-1.4 ... ]

I would prefer it way more than going with spdx-json, spdx-tv

@bardenstein what do you think?

bardenstein replied:
this is a bit different from other SBOM generators, that have things like --format=spdx-json or --format=cyclonedx-xml. 
But I'm fine with splitting them, having --encoding optional and defaulted to json.

houdini91 reviewed (commented):
Looks good i have no issue with us merging this as is getting 0.0.0 out in to the air .

Added Some comments, 

General comment:
I try and write down what i was thinking and see if people view this differently. 
In order to simplify the CLI UX i think its best we break our CLI in to two subcommand one to translate spdx and another for cdx.
This will allow simplify the way we  up with acutal `pFlag` default values allowing the user to run the simplist of command `go-cli cdx my_cdx.json` and `go-cli spdx my_spdx.json`, in turn this will simplify the usage help message to include said defaults for each subcommand.

For example the subcommand cdx can look similar to the following, note the defaults of the falgs reflect the opposing format latest supported format.

```

Translate between CycloneDX into SPDX SBOM, Bridging the gap between cyclonedx and spdx

Usage:
  protobom cdx [path] [flags]

Examples:

        protobom cdx sbom.cdx.json           output to default SPDX format.
	protobom cdx sbom.cdx.json -V 2.2    select SPDX specific version
	protobom cdx sbom.cdx.json -E text   select SPDX specific encoding
	protobom cdx sbom.cdx.json -o sbom.spdx.json       output to file
	protobom cdx --uri  text/spdx+text;version=2.2     select specific format URI
	

Flags:
  -c, --config string     Configuration file path
  -E, --encoding string   Select encoding, options=[text json] (default "json")

  -h, --help              help for protobom
  -D, --level string      Log depth level, options=[panic fatal error warning info debug trace]
  -o, --output string     Output path
      --uri string        Select uri, options=[text/spdx+text;version=2.3 text/spdx+json;version=2.3 text/spdx+text;version=2.2 text/spdx+json;version=2.2 application/vnd.cyclonedx+json;version=1.4 application/vnd.cyclonedx+json;version=1.5]
  -V, --spec-ver string   Select Specific version, options=[2.3 2.2] (default 2.3)
  -v, --verbose count     Log verbosity level [-v,--verbose=1] = info, [-vv,--verbose=2] = debug (default info)
      --version           version for protobom
```

Once the core library adds another format i would love for this usage to update once we pull in the new version (basing the content of the options and defaults from the underlining core library full format list as well as its default format list.

Another simplification i drop --encoding and --version and only support full uri including type,version and encoding in one string., and simplify default its value in to the underlining  core library default value.

manifestori commented:
> Looks good i have no issue with us merging this as is getting 0.0.0 out in to the air .
> 
> Added Some comments,
> 
> General comment: I try and write down what i was thinking and see if people view this differently. In order to simplify the CLI UX i think its best we break our CLI in to two subcommand one to translate spdx and another for cdx. This will allow simplify the way we up with acutal `pFlag` default values allowing the user to run the simplist of command `go-cli cdx my_cdx.json` and `go-cli spdx my_spdx.json`, in turn this will simplify the usage help message to include said defaults for each subcommand.
> 
> For example the subcommand cdx can look similar to the following, note the defaults of the falgs reflect the opposing format latest supported format.
> 
> ```
> 
> Translate between CycloneDX into SPDX SBOM, Bridging the gap between cyclonedx and spdx
> 
> Usage:
>   protobom cdx [path] [flags]
> 
> Examples:
> 
>         protobom cdx sbom.cdx.json           output to default SPDX format.
> 	protobom cdx sbom.cdx.json -V 2.2    select SPDX specific version
> 	protobom cdx sbom.cdx.json -E text   select SPDX specific encoding
> 	protobom cdx sbom.cdx.json -o sbom.spdx.json       output to file
> 	protobom cdx --uri  text/spdx+text;version=2.2     select specific format URI
> 	
> 
> Flags:
>   -c, --config string     Configuration file path
>   -E, --encoding string   Select encoding, options=[text json] (default "json")
> 
>   -h, --help              help for protobom
>   -D, --level string      Log depth level, options=[panic fatal error warning info debug trace]
>   -o, --output string     Output path
>       --uri string        Select uri, options=[text/spdx+text;version=2.3 text/spdx+json;version=2.3 text/spdx+text;version=2.2 text/spdx+json;version=2.2 application/vnd.cyclonedx+json;version=1.4 application/vnd.cyclonedx+json;version=1.5]
>   -V, --spec-ver string   Select Specific version, options=[2.3 2.2] (default 2.3)
>   -v, --verbose count     Log verbosity level [-v,--verbose=1] = info, [-vv,--verbose=2] = debug (default info)
>       --version           version for protobom
> ```
> 
> Once the core library adds another format i would love for this usage to update once we pull in the new library (basing the content of the options and defaults from the underlining core library full format list as well as its default format list.
> 
> Another simplification i drop --encoding and --version and only support full uri including type,version and encoding in one string., and simplify default its value in to the underlining core library default value. So something like
> 
> ```
> 
> Translate between CycloneDX into SPDX SBOM, Bridging the gap between cyclonedx and spdx
> 
> Usage:
>   protobom cdx [path] [flags]
> 
> Examples:
> 
>         protobom cdx sbom.cdx.json           output to default SPDX format.
> 	protobom cdx --uri  text/spdx+text;version=2.2     select specific format URI
> 	
> 
> Flags:
>   -c, --config string     Configuration file path
>   -h, --help              help for protobom
>   -D, --level string      Log depth level, options=[panic fatal error warning info debug trace]
>   -o, --output string     Output path
>       --uri string        Select uri, options=[text/spdx+text;version=2.3 text/spdx+json;version=2.3 text/spdx+text;version=2.2 text/spdx+json;version=2.2 application/vnd.cyclonedx+json;version=1.4 application/vnd.cyclonedx+json;version=1.5] (default text/spdx+json;version=2.3)
>   -v, --verbose count     Log verbosity level [-v,--verbose=1] = info, [-vv,--verbose=2] = debug (default info)
>       --version           version for protobom
> ```

You've raised several ideas and I would like to address all of them:

1) dedicated commands for cdx/spdx. I would argue that's less intuitive, as `cdx` is used alongside `cyclonedx` will cause confusions. then, renaming command to `sbom-convert cyclonedx sbom.spdx.json` is less ergonomic then `sbom-convert sbom.cdx.json`
Also note that I did not use a format as I'm utilizing the "reverse" functionality - detected format and choosing the reverse.
without it - `sbom-convert -f spdx sbom.cdx.json` 

In addition I removed the "convert" as it would be the default command. (`sbom-convert convert` == `sbom-convert`)

2) Supporting `uri` is a good idea, but some users just wants to `sbom-convert -f spdx sbom.cdx.json` and including the 
 long `uri` values is just a tiny bit more of a "power user" feature. So having us supports both, makes sense to me.

3) Sourcing supported formats, and encoding directly from `protobom` is a cool feature. It requires some string manipulations if we want to convert it to supported encoding/formats/versions which makes sense. however doing it programmatically is a bit overkill as we're not dealing with 5 new formats a month.
Even if we do supports this, we need `protobom` lib to expose a function or a map of supported `uri` we can pull from, then on a new release, we will need to upgrade `sbom-convert` with a PR anyway, I don't think its critical or cannot be resolved manually.

I do agree that sourcing CLI usage options from the same variable will make upgrade much easier, but this is also out of scope for this initial PR :D

puerco reviewed (commented):
Great work Ori, looks good to me in general :)

Sorry for so many comments, it is a big PR that touches a lot of areas. I have two concerns:

1. The PR is moving the translation API under the hood so it cannot be imported by other tools that may need a simple translation layer. I don't know if this is intentional, but I would like to have some lightweight translation API exposed from this module (see the diagram in the current README). It looks like your `convert` package fits the bill.
2. I feel we should shrink and simplify the image that gets built at release time. See my comments below.

Beyond those two, the rest of my comments are less important.

houdini91 reviewed (approved):
Lgtm

houdini91 commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM alpine:3.18@sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1
+
+ARG NAME=sbom-convert
+ENV NAME=${NAME}
+
+RUN apk add --no-cache \
+	bash \
+	docker-cli \
+	tini
+
+COPY scripts/entrypoint.sh /entrypoint.sh
+RUN chmod +x /entrypoint.sh
```
I also prefer to he a static binary image way.
It also helps with easier to create multi arch images..

I would say about ko. 
I prefer the default docker go releaser maybe minor but annoying issues using goreleaser with ko, it won't build the image unless you actually release (so if we run a snapshot only the binary is built) further more it will force build to have actual access to registry ..

@purco can you elaborate on the chaingaurd base image it looks interesting ..

I do like us using sboms on our own tool, we should discuss this I think.. what and how..

Fyi, Goreleaser it self  also can connect to a sbom topl directly to run the sboms on the artifacts.

manifestori replied:
Let's start with cleaning the dockerfile and then we can change to build with `ko`, I'll try changing the base alpine to cgr.dev/chainguard/static:sha***

manifestori commented on internal/convert/convert.go:
```diff
@@ -0,0 +1,43 @@
+package convert
```
Good point, I will move the convert back back to /pkg

manifestori commented on pkg/format/main.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+// ParseFormat parses the format string into a formats.Format
+func ParseFormat(fs string, encoding string) (*Format, error) {
```
Agreed, I kinda wrapped this for scoping purposes. I will open an issue on protobom so we can track this. 

We can remove this one once we have Formats complete

manifestori commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,28 @@
+    hooks:
+      - id: go-mod-tidy
+      - id: go-test-mod
+  - repo: https://github.com/alessandrojcm/commitlint-pre-commit-hook
+    rev: v9.5.0
+    hooks:
+      - id: commitlint
+        stages: [commit-msg]
+        additional_dependencies: ["@commitlint/config-conventional"]
```
Yeah, I'll disable that for now.

manifestori commented on .github/CODEOWNERS:
```diff
@@ -0,0 +1,54 @@
+# This is a comment.
```
Ill add us three as maintainers

manifestori commented on cmd/cli/options/errors.go:
```diff
@@ -0,0 +1 @@
+package options
```
Mostly a placeholder a d guidelines for next steps. I really want to have a proper errors lib. As we tend to neglect that and just create errors with random texts ;)

adefee reviewed (approved):
👍

manifestori commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,203 @@
+      - windows
+      - darwin
+    goarch:
+      - "386"
```
dropped

manifestori commented:
@houdini91 @puerco @bardenstein 

I have fixed most (almost all issues)
1) Docker image now 12mb instead of 59mb 
2) Removed some files
3) Improved build time by removing "386" builds
4) Added examples
5) Fixed default command, completion and version commands
6) CODEOWNERS
7) Moved `convert` from internal to pkg (so It can be used as a library)

**Missing**:
1) Options into commands (supported version etc), I'll add it on the next PR
2) Tests
3) URI support
4) updated Docs

houdini91 commented:
Awesome

jspeed-meyers commented:
I also support merging the current functionality (when others are good with it) and turning outstanding TODO's into GH issues. That makes it easier to have smaller, easier-to-understand PR changes.

Nice work!

puerco commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,28 @@
+    hooks:
+      - id: commitlint
+        stages: [commit-msg]
+        additional_dependencies: ["@commitlint/config-conventional"]
```
So, are we postponing this?

manifestori replied:
I apologize for missing that. I will temporarily disable the check and proceed with the merge.

manifestori replied:
Done :)

puerco reviewed (commented):
Looks good to me, I think the only change I'd request is to postpone the conventional commits checks. Otherwise I'd say we ship it :)

puerco reviewed (approved):
Awesome work @manifestori! lets ship it 🚀
